ts/ui/safe/safe.ts

```diff
@@ -0,0 +1,267 @@
+    src:                  'filterURL',
+    altimg:               'filterURL',
+    'class':              'filterClassList',
+    style:                'filterStyles',
+    id:                   'filterID',
+    fontsize:             'filterFontSize',
+    mathsize:             'filterFontSize',
+    scriptminsize:        'filterFontSize',
+    scriptsizemultiplier: 'filterSizeMultiplier',
+    scriptlevel:          'filterScriptLevel',
+    'data-':              'filterData'
+  };
+
+  /**
+   * The safe options from the document option list
+   */
+  public options: OptionList;
+
+  /**
+   * Shorthand for options.allow
+   */
+  public allow: OptionList;
+
+  /**
+   * The DOM adaptor from the document
+   */
+  public adaptor: DOMAdaptor<N, T, D>;
+
+  /**
+   * The methods for filtering the MathML attributes
+   */
+  public filterMethods: {[name: string]: FilterFunction<N, T, D>} = {
+    ...SafeMethods
+  };
+
+  /**
+   * @param {MathDocument<N,T,D>} document  The MathDocument we are sanitizing
+   * @param {OptionList} options            The safeOptions from the document
+   */
+  constructor(document: MathDocument<N, T, D>, options: OptionList) {
+    this.adaptor = document.adaptor;
+    this.options = options;
+    this.allow = this.options.allow;
+  }
+
+  /**
+   * Sanitize a MathItem's root MathML tree
+   *
+   * @param {MathItem<N,T,D>} math           The MathItem to sanitize
+   * @param {MathDocument<N,T,D>} document   The MathDocument in which it lives
+   */
+  public sanitize(math: MathItem<N, T, D>, document: MathDocument<N, T, D>) {
+    try {
+      math.root.walkTree(this.sanitizeNode.bind(this));
+    } catch (err) {
+      document.options.compileError(document, math, err);
+    }
+  }
+
+  /**
+   * Sanitize a node's attributes
+   *
+   * @param {MmlNode} node      The node to sanitize
+   */
+  protected sanitizeNode(node: MmlNode) {
+    const attributes = node.attributes.getAllAttributes();
+    for (const id of Object.keys(attributes)) {
+      if (this.filterAttributes.hasOwnProperty(id)) {
+        const value = this.filterMethods[this.filterAttributes[id]](this, attributes[id]);
+        if (value) {
+          if (value !== (typeof value === 'number' ? parseFloat(attributes[id] as string) : attributes[id])) {
+            attributes[id] = value;
+          }
+        } else {
+          delete attributes[id];
+        }
+      }
+    }
+  }
+
+  /**
+   * Sanitize a MathML input attribute
+   *
+   * @param {string} id      The name of the attribute
+   * @param {string} value   The value of the attribute
+   * @return {string|null}   The sanitized value
+   */
+  public mmlAttribute(id: string, value: string): string | null {
+    if (id === 'class') return null;
+    const filter = (this.filterAttributes.hasOwnProperty(id) ? this.filterAttributes[id] :
+                    (id.substr(0, 5) === 'data-' ? this.filterAttributes['data-'] : null));
+    if (!filter) return value;
```

Comment:
I thought we only allowed that for empty `return`.

Reply:
I didn't think so, but have changed it.